pkg/resolution/resolver/git/config.go

```diff
@@ -43,3 +51,54 @@ const (
 	// APISecretNamespaceKey is the config map key for the token secret's namespace
 	APISecretNamespaceKey = "api-token-secret-namespace"
 )
+
+type GitResolverConfig struct {
+	ScmTokens map[string]ScmInfo
+}
+
+type ScmInfo struct {
+	Timeout            string `json:"fetch-timeout"`
+	URL                string `json:"default-url"`
+	Revision           string `json:"default-revision"`
+	Org                string `json:"default-org"`
+	ServerURL          string `json:"server-url"`
+	SCMType            string `json:"scm-type"`
+	APISecretName      string `json:"api-token-secret-name"`
+	APISecretKey       string `json:"api-token-secret-key"`
+	APISecretNamespace string `json:"api-token-secret-namespace"`
+}
+
+func GetGitConfig(ctx context.Context) GitResolverConfig {
+	var scmInfo interface{} = &ScmInfo{}
+	structType := reflect.TypeOf(scmInfo).Elem()
+	gitResolverConfig := GitResolverConfig{ScmTokens: map[string]ScmInfo{}}
+	conf := framework.GetResolverConfigFromContext(ctx)
+	for k, v := range conf {
+		var tokenIdentifier, tokenKey string
+		key := strings.Split(k, ".")
+		if len(key) >= 3 && key[0] == "identifier" {
+			tokenKey = strings.Join(key[2:], ".")
+			tokenIdentifier = key[1]
+		} else {
+			tokenKey = k
+			tokenIdentifier = "default"
+		}
+		_, ok := gitResolverConfig.ScmTokens[tokenIdentifier]
+		if !ok {
+			gitResolverConfig.ScmTokens[tokenIdentifier] = ScmInfo{}
+		}
+		for i := range structType.NumField() {
+			field := structType.Field(i)
+			fieldName := field.Name
+			jsonTag := field.Tag.Get("json")
+			if tokenKey == jsonTag {
+				tokenDetails := gitResolverConfig.ScmTokens[tokenIdentifier]
+				var scm interface{} = &tokenDetails
+				structValue := reflect.ValueOf(scm).Elem()
```

Comment:
These three lines don't need to be within the loop over fields. You can get your `structValue` before starting the loop, update it for each field where the config name matches the json tag, and after the loop you can set it back to the `gitResolverConfig` object (so move line 99 after the loop.

Note that this implicitly let one set the same config multiple times, and the order of precedence depends on the order we get from`range conf` - we should either document this or check if a key is already set and fail.

Reply:
this is to handle the case like while parsing, we may receive different key, so i am parsing each key one by one

Reply:
Yes, I understand that, but you have two nested loops, the outer loop checks all the config keys, each with its own `tokenIdentifier`, the inner loop checks all known config keys against the `tokenKey` extracted from `tokenIdentifier` - so within the inner loop `tokenIdentifier` remains the same, until the next line of config is processed and the inner loop is executed again with a different `tokenIdentifier`.

The following should work:

```golang
	for k, v := range conf {
		var tokenIdentifier, tokenKey string
		key := strings.Split(k, ".")
		if len(key) >= 3 && key[0] == "identifier" {
			tokenKey = strings.Join(key[2:], ".")
			tokenIdentifier = key[1]
		} else {
			tokenKey = k
			tokenIdentifier = "default"
		}
		tokenDetails, ok := gitResolverConfig.ScmTokens[tokenIdentifier]
		if !ok {
			tokenDetails = ScmInfo{}
		}
		var scm interface{} = &tokenDetails
		structValue := reflect.ValueOf(scm).Elem()
		for i := range structType.NumField() {
			field := structType.Field(i)
			jsonTag := field.Tag.Get("json")
			if tokenKey == jsonTag {
				structValue.FieldByName(field.Name).SetString(v)
			}
		}
                gitResolverConfig.ScmTokens[tokenIdentifier] = structValue.Interface().(ScmInfo)
	}
```

Reply:
Actually, the way I have done is, it is doing all reflect thing only if the jsonTag is matching. If wrong key is provided then all those reflect code is not executed, which I think is better instead of doing it every time.

Reply:
NIT: I see, it's a small improvement in case of incorrect configuration, which should be the exception. I find the code I pasted above easier to parse, but that's probably just me 😅  No need to change it